src/LinuxTracing/LeafFunctionCallManager.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "LeafFunctionCallManager.h"
+
+#include <sys/mman.h>
+
+#include <vector>
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_linux_tracing {
+
+using orbit_grpc_protos::Callstack;
+
+// Let's unwind the stack from $rsp to $rbp using libunwindstack. If $rbp points to the current
+// frame, this will only include the locals and not the return address, so libunwindstack will
+// only be able to unwind one frame (the instruction pointer). If $rpb points to the previous frame,
+// as the leaf function does not contain frame pointers, the region from $rsp to $rbp will contain
+// the complete innermost frame, plus the locals of the caller's frame (but not the return address).
+// Therefore, libunwindstack will be able to compute two frames, where the outer one is the missing
+// caller and needs to be patched in.
+// As libunwindstack will try to unwind even further, unwinding errors will always be reported.
+// We need to fallback to plausibility checks to detect actual unwinding errors (such as whether the
+// frames are executable).
+// Note that we cannot simply set libunwindstack to unwind always two frames and compare the outer
+// frame with the respective one in the callchain carried by the perf_event_open event, as in case
+// of uprobes overriding the return addresses, both addresses would be identical even if the actual
+// addresses (after uprobe patching) are not.
+// More (internal) documentation on this in: go/stadia-orbit-leaf-frame-pointer
+Callstack::CallstackType LeafFunctionCallManager::PatchLeafFunctionCaller(
+    CallchainSamplePerfEvent* event, LibunwindstackMaps* current_maps,
+    LibunwindstackUnwinder* unwinder) {
+  CHECK(event != nullptr);
+  CHECK(current_maps != nullptr);
+  CHECK(unwinder != nullptr);
+  CHECK(event->GetCallchainSize() > 2);
+
+  uint64_t stack_size =
```

Comment:
So related to your answer to the other comment, could this be negative (in theory, it's a `uint64_t`)? Or did I misunderstand? How does `Unwind` react to a then very big number?

Reply:
That's a good point, I explicitly handled the case that rbp is smaller than rsp now and abort early.